drivers/nvme.cc

```diff
@@ -0,0 +1,693 @@
+    nvme_cq_entry_t* cqbuf = (nvme_cq_entry_t*) alloc_phys_contiguous_aligned(qsize * sizeof(nvme_cq_entry_t),4096);
+    
+    nvme_adminq_attr_t aqa;
+    aqa.val = 0;
+    aqa.asqs = aqa.acqs = qsize - 1;
+    
+    u32* sq_doorbell = _control_reg->sq0tdbl;
+    u32* cq_doorbell = (u32*) ((u64)sq_doorbell + _doorbellstride);
+
+    _admin_queue = std::unique_ptr<nvme_admin_queue_pair>(new nvme_admin_queue_pair(_id,0, qsize, _dev, sqbuf, sq_doorbell, cqbuf, cq_doorbell, _ns_data));
+    
+    register_admin_interrupts();
+    
+    mmio_setl(&_control_reg->aqa, aqa.val);
+    mmio_setq(&_control_reg->asq, (u64) mmu::virt_to_phys((void*) sqbuf));
+    mmio_setq(&_control_reg->acq, (u64) mmu::virt_to_phys((void*) cqbuf));
+}
+
+int nvme::create_io_queue(int qsize, int qprio) 
+{
+    u32* sq_doorbell;
+    u32* cq_doorbell;
+    int id = _io_queues.size() + 1;
+    int iv = id;
+    qsize = (qsize < _control_reg->cap.mqes) ? qsize : _control_reg->cap.mqes + 1;
+
+    nvme_sq_entry_t* sqbuf = (nvme_sq_entry_t*) alloc_phys_contiguous_aligned(qsize * sizeof(nvme_sq_entry_t),4096);
+    nvme_cq_entry_t* cqbuf = (nvme_cq_entry_t*) alloc_phys_contiguous_aligned(qsize * sizeof(nvme_cq_entry_t),4096);
+    assert(sqbuf);
+    assert(cqbuf);
+    memset(sqbuf,0,sizeof(nvme_sq_entry_t)*qsize);
+    memset(cqbuf,0,sizeof(nvme_cq_entry_t)*qsize);
+
+    // create completion queue
+    nvme_acmd_create_cq_t* cmd = (nvme_acmd_create_cq_t*) malloc(sizeof(nvme_acmd_create_cq_t));
+    assert(cmd);
+    memset(cmd, 0, sizeof (*cmd));
+    
+    cmd->qid = id;
+    cmd->qsize = qsize - 1;
+    cmd->iv = iv;
+    cmd->pc = 1;
+    cmd->ien = 1;
+    cmd->common.opc = NVME_ACMD_CREATE_CQ;
+    cmd->common.prp1 = (u64) mmu::virt_to_phys(cqbuf);
+
+    // create submission queue
+    nvme_acmd_create_sq_t* cmd_sq = (nvme_acmd_create_sq_t*) malloc(sizeof(nvme_acmd_create_sq_t));
+    assert(cmd_sq);
+    memset(cmd_sq, 0, sizeof(nvme_acmd_create_sq_t));
+    
+    cmd_sq->pc = 1;
+    cmd_sq->qprio = qprio; // 0=urgent 1=high 2=medium 3=low
+    cmd_sq->qid = id;
+    cmd_sq->cqid = id;
+    cmd_sq->qsize = qsize - 1;
+    cmd_sq->common.opc = NVME_ACMD_CREATE_SQ;
+    cmd_sq->common.prp1 = (u64) mmu::virt_to_phys(sqbuf);
+
+    sq_doorbell = (u32*) ((u64) _control_reg->sq0tdbl + 2 * _doorbellstride * id);
+    cq_doorbell = (u32*) ((u64) sq_doorbell + _doorbellstride);
+
+    _io_queues.push_back(std::unique_ptr<nvme_io_queue_pair>(new nvme_io_queue_pair(_id, iv, qsize, _dev, sqbuf, sq_doorbell, cqbuf, cq_doorbell, _ns_data)));
+    
+    register_interrupt(iv,id-1);
+
+    _admin_queue->submit_and_return_on_completion(std::unique_ptr<nvme_sq_entry_t>((nvme_sq_entry_t*)cmd));
+    _admin_queue->submit_and_return_on_completion(std::unique_ptr<nvme_sq_entry_t>((nvme_sq_entry_t*)cmd_sq));
+
+    return id -1;
+}
+
+void nvme::create_io_queues_foreach_cpu()
+{
```

Comment:
I see here a potential for a lot of code re-use between `create_io_queue()` and `create_io_queues_foreach_cpu()`. Can we have some helper method to implement common logic?